### PR TITLE
distsqlrun: Fix MergeJoiner correctness issue with NULL values

### PR DESCRIPTION
**Pull request by lgo**
Previously, when a distSQL merge join was being executed the comparison
check was using Datum.Compare while processing the incoming streams we
are joining on.

This is problematic when both sides of the stream have a NULL value,
because this comparison is considered equal which is a violation of the
semantics of equals (=) with SQL NULLs.

Unfortunately, this bug was present before distSQL merge joins were
being planned (#17214) and so it exists in 1.1.0.

### CODE REVIEW

**Comment by lego** [comment]
On a side note: this behavior we might like for an anti join, and in particular the scrub checks I've been working on ;). In the (near) future I expect I'll plumb a flag into here in order to make NULL comparisons truth-y.

File: pkg/sql/distsqlrun/stream_merger.go [hunk 1]
```diff
@@ -107,6 +107,13 @@ func CompareEncDatumRowForMerge(
 		lIdx := ord.ColIdx
 		rIdx := rightOrdering[i].ColIdx
 		cmp, err := lhs[lIdx].Compare(&lhsTypes[lIdx], da, evalCtx, &rhs[rIdx])
+		// If the datums are both DNull, we need to follow SQL semantics
+		// where they are not equal. This differs from our datum semantics
+		// where they are equal.
+		if lhs[lIdx].Datum == tree.DNull && rhs[rIdx].Datum == tree.DNull {
+			// We can return either -1 or 1, it does not change the behavior.
+			return -1, nil
+		}
```
Comment (vivekmenezes) [inline pkg/sql/distsqlrun/stream_merger.go]: perhaps move this to before we call Compare()

Comment (lego) [reply pkg/sql/distsqlrun/stream_merger.go]: I should have mentioned this in the comment, but Compare is on two EncDatums so we don't yet know if the Datums are decoded (I'm assuming), so this happens specifically after.

**Comment by lego** [comment]
Agh, the tests failed. The failure doesn't seem to be deterministic as they were working earlier, which is a bit worrisome. I'll look at it in the morning.

**Comment by RaduBerinde** [comment]
Please file an issue to push down `NOT NULL` constraints on equality columns during planning (depending on join type). This can reduce the scan ranges.

---

Review status: 0 of 2 files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---

*[pkg/sql/distsqlrun/stream_merger.go, line 116 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/20090#-Kz1aEywOoiXpLpPFJdr-r2-116:-Kz3dbhoKbrtGpINCsdP:b-dzgolh) ([raw file](https://github.com/cockroachdb/cockroach/blob/263c1edaa3169fa57eb10253c9fcaaf07c4fee64/pkg/sql/distsqlrun/stream_merger.go#L116)):*
<details><summary><i>Previously, lego (Joey Pereira) wrote…</i></summary><blockquote>

I should have mentioned this in the comment, but Compare is on two EncDatums so we don't yet know if the Datums are decoded (I'm assuming), so this happens specifically after.
</blockquote></details>

Compare doesn't necessarily decode them (it could compare key encodings directly). You can use `EncDatum.IsNull`.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20090#-:-Kz3dyQUqejqcyXym3Re:b-uei7dy)*

**Comment by lego** [comment]
Ah, I wasn't sure what you were referring to previously. The issue is referenced above this comment.

---

Review status: 0 of 2 files reviewed at latest revision, 1 unresolved discussion.

---

*[pkg/sql/distsqlrun/stream_merger.go, line 116 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/20090#-Kz1aEywOoiXpLpPFJdr-r2-116:-Kz4awoPxeN2x4szcXUN:b-emfdxo) ([raw file](https://github.com/cockroachdb/cockroach/blob/263c1edaa3169fa57eb10253c9fcaaf07c4fee64/pkg/sql/distsqlrun/stream_merger.go#L116)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

Compare doesn't necessarily decode them (it could compare key encodings directly). You can use `EncDatum.IsNull`.
</blockquote></details>

Awesome, also moved this above the `Compare()`.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20090#-:-Kz4ar3n6NxquhLYDt6z:bquxx6a)*

**Comment by lego** [comment]
It seems like `algebraicSetOp` also uses the compare function that is being changed, but the code path that reaches it is never used (`algebraicSetOp.Run()` -> `algebraicSetOp.exceptAll()` -> `CompareEncDatumRowForMerge`). I'm guesisng this isn't an issue as it's not used, @RaduBerinde?

**Comment by RaduBerinde** [comment]
I'm not sure how algebraicSetOp is supposed to work with NULLs, CC @arjunravinarayan

**Comment by lego** [comment]
Review status: 0 of 3 files reviewed at latest revision, 2 unresolved discussions.

---

*[pkg/sql/distsqlrun/mergejoiner_test.go, line 359 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/20090#-Kz5CupqYWLD3kulyWZC:-Kz5CuprVRTrZFx7XEEG:b-cj7qpb) ([raw file](https://github.com/cockroachdb/cockroach/blob/69d9a4091a30702aaf4147c2ef5e0e04918d1692/pkg/sql/distsqlrun/mergejoiner_test.go#L359)):*
> ```Go
> 			},
> 		},
> 		{
> ```

I added these tests which should test this problem, but they seem to pass prior to the fix.

As far as I know, this should still be using the same logic in from the streamMerger, so I'm not sure why it is still working with the bug. Could it maybe be how these `null` are interpreted? I dug a bit but didn't find an answer. @RaduBerinde would you happen to know why?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20090)*

**Comment by RaduBerinde** [comment]
*[pkg/sql/distsqlrun/mergejoiner_test.go, line 359 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/20090#-Kz5CupqYWLD3kulyWZC:-Kz5FakaQPjgteR11ArA:b-q1me9h) ([raw file](https://github.com/cockroachdb/cockroach/blob/69d9a4091a30702aaf4147c2ef5e0e04918d1692/pkg/sql/distsqlrun/mergejoiner_test.go#L359)):*
<details><summary><i>Previously, lego (Joey Pereira) wrote…</i></summary><blockquote>

I added these tests which should test this problem, but they seem to pass prior to the fix.

As far as I know, this should still be using the same logic in from the streamMerger, so I'm not sure why it is still working with the bug. Could it maybe be how these `null` are interpreted? I dug a bit but didn't find an answer. @RaduBerinde would you happen to know why?
</blockquote></details>

Hm, I don't know why that would be the case.. I would dig deeper, try to print out the result of every compare that happens in the stream merger

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20090)*

**Comment by lego** [comment]
Review status: 0 of 3 files reviewed at latest revision, 2 unresolved discussions.

---

*[pkg/sql/distsqlrun/mergejoiner_test.go, line 359 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/20090#-Kz5CupqYWLD3kulyWZC:-KzP4dFrsb_3v8XUnc27:btbdoge) ([raw file](https://github.com/cockroachdb/cockroach/blob/69d9a4091a30702aaf4147c2ef5e0e04918d1692/pkg/sql/distsqlrun/mergejoiner_test.go#L359)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

Hm, I don't know why that would be the case.. I would dig deeper, try to print out the result of every compare that happens in the stream merger
</blockquote></details>

Pinned it down. It was because the tests had an `OnExpr` checking equality while in the code the expression was implicitly removed when the equalities are parsed. This caused the test to filter out the bad results. The tests now correctly test the bug :)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20090)*

File: pkg/sql/logictest/testdata/logic_test/distsql_join [hunk 2]
```diff
@@ -194,3 +194,48 @@ query T
...
+
+query II rowsort
+SELECT l.k, r.k FROM (SELECT * FROM distsql_mj_test ORDER BY k) l INNER JOIN (SELECT * FROM distsql_mj_test ORDER BY k) r ON l.k = r.k
+----
+
+# If NULL = NULL is joined, we should expect a row of (NULL, NULL)
+query T
+SELECT "URL" FROM [EXPLAIN (DISTSQL) (SELECT l.k, r.k FROM (SELECT * FROM distsql_mj_test ORDER BY k) l INNER JOIN (SELECT * FROM distsql_mj_test ORDER BY k) r ON l.k = r.k)]
+----
+https://cockroachdb.github.io/distsqlplan/decode.html?eJzEkkFr4zAQhe_7K5Y57RIVIifOwVDwNYUmJe2thOBaU1fF8bijMbSE_Pdi65DE1EpzKL1Jo_nemydmBxUZXGRbdJA8ggYFMawV1Ew5Okfcln3T3LxDMlZgq7qRtrxWkBMjJDsQKyVCAgu6ohoUGJTMll3TXgE1ckCcZAVCMturI1kdln3InkpcYWaQT8ShZrvN-CM11ol7Kzfb142gE1CwbCT5m2oY8teX-N8TS9861aNB8eiXw01-Mtx0UPyg2VTEBhlNfxfOt3wx4S1ygTdkq_6YJT7Lv1SP_l-zLV788fA9Ko0GQ8QnIc5s9QpdTZXDby32uE2ApkD_I44azvGOKe9s_HXZcV3BoBP_OvOXeeWf2gGPYR2EozAcBeH4BNZ9eBKEp2Hn6QXOUR-Og_C457ze__kMAAD__4m9mGY=
```
Comment (vivekmenezes) [inline pkg/sql/logictest/testdata/logic_test/distsql_join]: Why do you need the SELECT "URL" ?

Comment (lego) [reply pkg/sql/logictest/testdata/logic_test/distsql_join]: A pretty unconventional way to assert the plan is using a merge join.

File: pkg/sql/logictest/testdata/logic_test/distsql_join [hunk 3]
```diff
@@ -194,3 +194,48 @@ query T
...
+DELETE FROM distsql_mj_test WHERE TRUE;
+
+statement ok
+INSERT INTO distsql_mj_test VALUES (NULL)
+
+query II rowsort
+SELECT l.k, r.k FROM (SELECT * FROM distsql_mj_test ORDER BY k) l INNER JOIN (SELECT * FROM distsql_mj_test ORDER BY k) r ON l.k = r.k
+----
+
+# If NULL = NULL is joined, we should expect a row of (NULL, NULL)
```
Comment (vivekmenezes) [inline pkg/sql/logictest/testdata/logic_test/distsql_join]: Can you place this comment above the previous query. Should it say, "we shouldn't expect a row of (NULL, NULL)"

Comment (lego) [reply pkg/sql/logictest/testdata/logic_test/distsql_join]: Oof, thanks for catching that mistake.

**Comment by vivekmenezes** [comment]
:lgtm:

---

Review status: 0 of 3 files reviewed at latest revision, 4 unresolved discussions, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20090#-:-KzPMN3w6Jou1vBoWcDw:bnfp4nl)*

**Comment by vivekmenezes** [comment]
radu is on vacation so please feel free to merge

---

Review status: 0 of 3 files reviewed at latest revision, 4 unresolved discussions, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20090#-:-KzPMP294fX01MyZREZh:b-issy10)*

**Comment by lego** [comment]
TYFR! :) (the last change was just a release note update, btw)

**Comment by RaduBerinde** [comment]
LGTM, thanks.

**Comment by arjunravinarayan** [comment]
Sorry for the delay, I was out of office.

@lego: the code path is not used because we currently don't plan set operations. However we still want the semantics to be right.

I did not think seriously about the treatment of NULLs in algebraic set op, and cannot find documentation about how Postgres handles them either. I'm flagging this for followup when we do algebraic set op planning in #10432. Thank you for pointing that out @raduberinde.

---







*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20090#-:-KzzMO7pWol8eoIO60QH:bmoqozh)*

**Comment by arjunravinarayan** [comment]
I did some spelunking, and [Markus Winand](http://modern-sql.com/concept/null#equals-vs-distinct) points out that NULLs in group operations are supposed to sort into a single group, and that the SQL:2016 standard has a longer section on this, that we should ensure we are in compliance with (SQL:2016-2: §4.1.5 and SQL:2016-2: §9.12).

**Comment by arjunravinarayan** [comment]
I apologize, I posted in the wrong issue. Deleting here and reposting there so that your email client registers my apology (then deleting this apology as well).

**Pull request merged**